network-config-analyzer/NetworkConfigQuery.py

```diff
@@ -537,90 +543,250 @@ def pretty_print_diff(diff_list):
-            explanation += f'Removed connections:\n{self.pretty_print_diff(removed)}\n'
-
-            return QueryAnswer(False, 'Base branch and analyzed commit are not semantically equivalent.', explanation)
-        return QueryAnswer(True, 'Base branch and analyzed commit are semantically equivalent.')
+                    all_diff[key].append(
+                        SemanticDiffQuery.SingleDiff(pod1, pod2, old_conns - new_conns, new_conns - old_conns))
+
+        """
+        3.2  lost connections between intersected peers and removed ipBlock peers
+        3.3  lost/new connections between intersected peers and intersected ipBlock peers due to changes in policies and labels of pods/namespaces
+        3.4  new connections between intersected peers and added ipBlock peers
+        """
+        key = 'Changed connections between persistent peers and ipBlock peers'
+        all_diff[key] = []
+        peers = captured_pods | disjoint_ip_blocks
+        for pod1 in peers:
+            for pod2 in disjoint_ip_blocks if pod1 in captured_pods else captured_pods:
+                if pod1 == pod2:
+                    continue
+                _, old_conns, _ = self.config1.allowed_connections(pod1, pod2)
+                _, new_conns, _ = self.config2.allowed_connections(pod1, pod2)
+                if new_conns != old_conns:
+                    all_diff[key].append(
+                        SemanticDiffQuery.SingleDiff(pod1, pod2, old_conns - new_conns, new_conns - old_conns))
+
+        # 4.1. new connections between intersected peers and added peers
+        key = 'New connections between persistent peers and added peers'
+        all_diff[key] = []
+        for pair in itertools.product(intersected_peers, added_peers):
+            _, new_conns, _ = self.config2.allowed_connections(pair[0], pair[1])
+            if str(new_conns) != 'No connections':
+                all_diff[key].append(
+                    SemanticDiffQuery.SingleDiff(pair[0], pair[1], None, new_conns))
+
+            _, new_conns, _ = self.config2.allowed_connections(pair[1], pair[0])
+            if str(new_conns) != 'No connections':
+                all_diff[key].append(
+                    SemanticDiffQuery.SingleDiff(pair[1], pair[0], None, new_conns))
+
+        # 5.1. new connections between added peers
+        key = 'New connections between added peers'
+        all_diff[key] = []
+        for pair in itertools.permutations(added_peers, 2):
+            _, new_conns, _ = self.config2.allowed_connections(pair[0], pair[1])
+            if str(new_conns) != 'No connections':
+                all_diff[key].append(
+                    SemanticDiffQuery.SingleDiff(pair[0], pair[1], None, new_conns))
+
+        # 5.2. new connections between added peers and new ipBlock peers
+        key = 'New connections between added peers and new ipBlock peers'
+        all_diff[key] = []
+        for pair in itertools.product(added_peers, disjoint_ip_blocks):
```

Comment:
```suggestion
        for pair in itertools.product(added_peers, self.config2.get_referenced_ip_blocks()):
```

Reply:
No. same as for 1.2 above

Reply:
Similarly to 1.2, what we really need here is using the result of applying disjoint_ip_blocks() to self.config2.get_referenced_ip_blocks() and to IpBlock.get_all_ips_block().

Reply:
good catch, fixed - thanks.